frost-secp256k1-tr/src/lib.rs

```diff
@@ -0,0 +1,628 @@
+    /// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-2.2.2.3
+    fn H3(m: &[u8]) -> <<Self::Group as Group>::Field as Field>::Scalar {
+        hash_to_scalar((CONTEXT_STRING.to_owned() + "nonce").as_bytes(), m)
+    }
+
+    /// H4 for FROST(secp256k1, SHA-256)
+    ///
+    /// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-2.2.2.4
+    fn H4(m: &[u8]) -> Self::HashOutput {
+        hash_to_array(&[CONTEXT_STRING.as_bytes(), b"msg", m])
+    }
+
+    /// H5 for FROST(secp256k1, SHA-256)
+    ///
+    /// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-2.2.2.5
+    fn H5(m: &[u8]) -> Self::HashOutput {
+        hash_to_array(&[CONTEXT_STRING.as_bytes(), b"com", m])
+    }
+
+    /// HDKG for FROST(secp256k1, SHA-256)
+    fn HDKG(m: &[u8]) -> Option<<<Self::Group as Group>::Field as Field>::Scalar> {
+        Some(hash_to_scalar(
+            (CONTEXT_STRING.to_owned() + "dkg").as_bytes(),
+            m,
+        ))
+    }
+
+    /// HID for FROST(secp256k1, SHA-256)
+    fn HID(m: &[u8]) -> Option<<<Self::Group as Group>::Field as Field>::Scalar> {
+        Some(hash_to_scalar(
+            (CONTEXT_STRING.to_owned() + "id").as_bytes(),
+            m,
+        ))
+    }
+
+    /// Generates the challenge as is required for Schnorr signatures.
+    fn challenge(R: &Element<S>, verifying_key: &VerifyingKey, msg: &[u8]) -> Challenge<S>
+    {
+        let mut preimage = vec![];
+        let tweaked_public_key = tweaked_public_key(&verifying_key.to_element(), &[]);
+        preimage.extend_from_slice(&R.to_affine().x());
+        preimage.extend_from_slice(&tweaked_public_key.to_affine().x());
+        preimage.extend_from_slice(msg);
+        Challenge::from_scalar(S::H2(&preimage[..]))
+    }
+
+    /// determine tweak is need
+    fn is_need_tweaking() -> bool {
+        true
+    }
+
+    /// aggregate tweak z
+    fn aggregate_tweak_z(
+        z: <<Self::Group as Group>::Field as Field>::Scalar,
+        challenge: &Challenge<S>,
+        verifying_key: &Element<S>,
+    ) -> <<Self::Group as Group>::Field as Field>::Scalar
+    {
+        let t = tweak(&verifying_key, &[]);
+        z + t * challenge.clone().to_scalar()
+    }
+
+    /// compute tweaked signature_share
+    fn compute_tweaked_signature_share(
+        signer_nonces: &round1::SigningNonces,
+        binding_factor: frost::BindingFactor<S>,
+        group_commitment: frost_core::GroupCommitment<S>,
+        lambda_i: <<Self::Group as Group>::Field as Field>::Scalar,
+        key_package: &frost::keys::KeyPackage<S>,
+        challenge: Challenge<S>,
+    ) -> round2::SignatureShare
+    {
+        let mut sn = signer_nonces.clone();
+        if group_commitment.y_is_odd() {
+            sn.negate_nonces();
+        }
+
+        let mut kp = key_package.clone();
+        if key_package.verifying_key().y_is_odd() {
+            kp.negate_signing_share();
+        }
```

Comment:
IIUC this should also use the y_is_odd of the tweaked public key!

```rust
let tweaked_pubkey = tweaked_public_key(key_package.verifying_key().element(), &[]);
        if tweaked_pubkey.to_affine().y_is_odd().into() {
            kp.negate_signing_share();
        }
```

Reply:
or this might be it:

```rust
let public_key = key_package.verifying_key();
let pubkey_is_odd = public_key.y_is_odd();
let tweaked_pubkey_is_odd = tweaked_public_key(publickey.element(), &[]).is_odd().into();
if pubkey_is_odd != tweaked_pubkey_is_odd {
  kp.negate_signing_share();
}
```